From 9517b3173af961ea66721cfc48cd47e50a704388 Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Wed, 4 Nov 2015 17:17:06 +0800
Subject: [PATCH] Work around the failure of ExitBootServices()

ExitBootServices() will fail if any of the event handlers change
the memory map. In which case, we must be prepared to retry, but
only once so that we're guaranteed to exit on repeated failures
instead of spinning forever. This fix refers to the workaround
made by Linux kernel.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 grub-core/kern/efi/mm.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/grub-core/kern/efi/mm.c b/grub-core/kern/efi/mm.c
index 461deb0..7620a47 100644
--- a/grub-core/kern/efi/mm.c
+++ b/grub-core/kern/efi/mm.c
@@ -158,6 +158,7 @@ grub_efi_finish_boot_services (grub_efi_uintn_t *outbuf_size, void *outbuf,
 {
   grub_efi_boot_services_t *b;
   grub_efi_status_t status;
+  int called_exit = 0;
 
 #if defined (__i386__) || defined (__x86_64__)
   const grub_uint16_t apple[] = { 'A', 'p', 'p', 'l', 'e' };
@@ -167,6 +168,7 @@ grub_efi_finish_boot_services (grub_efi_uintn_t *outbuf_size, void *outbuf,
 			   apple, sizeof (apple)) == 0);
 #endif
 
+get_mem_map:
   if (grub_efi_get_memory_map (&finish_mmap_size, finish_mmap_buf, &finish_key,
 			       &finish_desc_size, &finish_desc_version) < 0)
     return grub_error (GRUB_ERR_IO, "couldn't retrieve memory map");
@@ -186,7 +188,21 @@ grub_efi_finish_boot_services (grub_efi_uintn_t *outbuf_size, void *outbuf,
   status = efi_call_2 (b->exit_boot_services, grub_efi_image_handle,
 		       finish_key);
   if (status != GRUB_EFI_SUCCESS)
-    return grub_error (GRUB_ERR_IO, "couldn't terminate EFI services");
+    {
+      /*
+       * ExitBootServices() will fail if any of the event
+       * handlers change the memory map. In which case, we
+       * must be prepared to retry, but only once so that
+       * we're guaranteed to exit on repeated failures instead
+       * of spinning forever.
+       */
+      if (called_exit)
+        return grub_error (GRUB_ERR_IO, "couldn't terminate EFI services");
+
+      called_exit = 1;
+      grub_free (finish_mmap_buf);
+      goto get_mem_map;
+    }
 
   grub_efi_is_finished = 1;
   if (outbuf_size)
-- 
1.9.1

